src/nrncvode/sptree.hpp

```diff
@@ -13,123 +13,117 @@
 
 #define STRCMP(a, b) (a - b)
 
-template <typename SPBLK>
-struct SPTREE {
-    SPBLK* root; /* root node */
-
-    int enqcmps; /* compares in spenq */
+template <typename T>
+class SPTree {
+  public:
+    SPTree() = default;
+
+    // Is this SPTree empty.
+    bool empty() const;
+
+    // Return the value of enqcmps.
+    int get_enqcmps() const;
+
+    // Insert item in a tree.
+    //
+    // Put n after all other nodes with the same key; when this is
+    // done, n will be the root of the splay tree, all nodes
+    // with keys less than or equal to that of n will be in the
+    // left subtree, all with greater keys will be in the right subtree;
+    // the tree is split into these subtrees from the top down, with rotations
+    // performed along the way to shorten the left branch of the right subtree
+    // and the right branch of the left subtree
+    void enqueue(T* n);
+
+    // Return and remove first node.
+    //
+    // Remove and return the first node; this deletes
+    // (and returns) the leftmost node, replacing it with its right
+    // subtree (if there is one); on the way to the leftmost node, rotations
+    // are performed to shorten the left branch of the tree.
+    T* dequeue();
+
+    // Return the first element in the tree.
+    //
+    // Returns a reference to the first item;
+    // event, and the old left branch is shortened, as if the tree had
+    // been splayed about the first element; this is done by dequeueing
+    // the head and then making the resulting queue the right son of
+    // the head returned by spdeq.
+    T* first();
+
+    // Remove node from a tree.
+    //
+    // n is removed; the resulting splay tree has been splayed
+    // around its new root, which is the successor of n
+    void remove(T* n);
+
+    // Given key, find a node in a tree.
+    //
+    // Splays the found node to the root.
+    T* find(double key);
+
+    // Apply a function to nodes in ascending order.
+    //
+    // If n is given, start at that node, otherwise start from the head.
+    void apply_all(void (*f)(const T*, int), T* n) const;
+
+  private:
```

Comment:
@1uc there is a private here.

Reply:
Yes, I saw. The question is can anything from above that line be moved below the line? If the answer is yes, consider doing so. If the answer is no: well done.